pkg/sql/sqlinstance/instancestorage/instancereader.go

```diff
@@ -121,6 +123,43 @@ func (r *Reader) WaitForStarted(ctx context.Context) error {
 	}
 }
 
+// GetAllInstancesUsingTxn reads all instances using the given transaction and returns
+// live instances only.
+func (r *Reader) GetAllInstancesUsingTxn(
+	ctx context.Context, tenantCodec keys.SQLCodec, tableID catid.DescID, txn *kv.Txn,
+) ([]sqlinstance.InstanceInfo, error) {
+	instancesTablePrefix := tenantCodec.TablePrefix(uint32(tableID))
+	rows, err := txn.Scan(ctx, instancesTablePrefix, instancesTablePrefix.PrefixEnd(), 0 /* maxRows */)
+	if err != nil {
+		return nil, err
+	}
+	rowcodec := makeRowCodec(tenantCodec, tableID)
+	decodedRows := make([]instancerow, 0, len(rows))
+	for _, row := range rows {
+		decodedRow, err := rowcodec.decodeRow(row.Key, row.Value)
+		if err != nil {
+			return nil, err
+		}
+		decodedRows = append(decodedRows, decodedRow)
+	}
+	filteredRows, err := r.getAllUniqueLiveInstances(ctx, decodedRows)
+	if err != nil {
+		return nil, err
+	}
+	instances := make([]sqlinstance.InstanceInfo, 0, len(rows))
+	for _, row := range filteredRows {
+		instanceInfo := sqlinstance.InstanceInfo{
+			InstanceID:      row.instanceID,
+			InstanceRPCAddr: row.rpcAddr,
+			InstanceSQLAddr: row.sqlAddr,
+			SessionID:       row.sessionID,
+			Locality:        row.locality,
+		}
+		instances = append(instances, instanceInfo)
+	}
+	return instances, nil
```

Comment:
nit: extract a helper to convert the `[]instancerow` into `[]sqlinstance.InstanceInfo`. I'd probably also extract a constructor for a single InstanceInfo from a row like `makeInstanceInfo(row instancerow) sqlinstance.InstanceInfo` so that is' like:

```go
func makeInstanceInfos(rows []instancerow) []sqlinstance.InstanceInfo {
    ret := make([]sqlinstance.InstanceInfo, len(rows))
    for i := range rows {
        ret[i] = makeInstanceInfo(rows[i])
    }
    return ret
}
```